c7n/resources/s3.py

```diff
@@ -1171,26 +1171,33 @@ class ToggleVersioning(BucketActionBase):
         enabled={'type': 'boolean'})
     permissions = ("s3:PutBucketVersioning",)
 
+    def process_versioning(self, resource, state):
+        client = bucket_client(
```

Comment:
session creation should be hoisted up, and passed as a parameter here instead of being created per resource

Reply:
bucket_client requires a resource as a parameter to get the region of the bucket, here we would actually be creating less sessions as we're only creating sessions for buckets that will be acted on instead of all buckets previously